Geraete

hoffmatteo opened the pull request:
### Fertige erste Version des Geräte-Fragments:

- Geräte in einer Recyclerview darstellen mit Optionen zum Sortieren (z.b Top-Verbraucher) --> Anzeige durch Unterstreichen
- Klick auf ein Gerät führt zur Edit-Seite
- Klick auf ActionButton führt zur Create-Seite 
- Edit und Create Seite haben Spinner, die automatisch die Kategorien und Räume erfassen
- Automatische Berechnung des Jahresverbrauch in kWh/Jahr anhand Volllast-/StandBy-Verbrauch und Betriebszeit (Volllast)
- Datenbank wurde vollständig angebunden und um Jahresverbrauch erweitert
- Tobias hat noch ein ActionButtonMenu und Create/Edit für die Produzenten-Geräte hinzugefügt


### TODO:
- Zweite Recyclerview für Produzenten (Absprache mit Tobias)
- Icons der Kategorien in Textview einbinden

hoffmatteo committed:
Created recyclerview + adapter, linked it with database

hoffmatteo committed:
Added create geraet, contains error with double parsing

hoffmatteo committed:
Finished add gerät

hoffmatteo committed:
Added edit/delete Fragment with alertdialog and database connection

hoffmatteo committed:
Better empty string check, created sorting-bar to sort geraete, verbrauch is now shown in list, fixed jahresverbrauch

tobireith committed:
Implementierung der des FAB Menus in Geräte zur Auswahl ob das neue Gerät Strom erzeugt oder verbraucht.

hoffmatteo committed:
Connected Haushalt-Spinner in MainActivity to database

tobireith committed:
Nun kann man Produzenten hinzufügen. Die DB wurde so geändert, dass der Eigenverbrauch für Produzenten gespeichert wird.

tobireith committed:
Man kann nun auch bestehende Produzenten bearbeiten.

Hierzu gibt es nun ein extra Fragment mit Layout.

Außerdem wurden noch die Buttons gefixt, damit diese nicht nach oben springen sobald Man Texte eingibt

hoffmatteo committed:
Added new select for verbraucher, added sorting with raum + name, fixed spinner in geraete new/edit/del

tobireith committed:
Select Funktion für getAllProduzenten wurde hinzugefügt

tobireith committed:
Merge branch 'geraete' into geraete_produzent

# Conflicts:
#	app/src/main/java/com/example/stromtracker/database/DataRepository.kt
#	app/src/main/java/com/example/stromtracker/database/GeraeteDAO.kt
#	app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt
#	app/src/main/res/layout/fragment_geraete.xml

tobireith committed:
Fixed Select Befehl für Produzenten und den Spinner ItemSelected Listener

tobireith committed:
Merge pull request #18 from emuguy1/geraete_produzent

Geraete Produzent into Geraete - Funktionalität und DB-Erweiterung

hoffmatteo committed:
Added back underline when sorted, fixed dimensions, added strings

hoffmatteo committed:
String fixes v2

hoffmatteo committed:
Fixed Zeit-Textview for smaller screens

emuguy1 commented on app/src/main/java/com/example/stromtracker/MainActivity.kt:
```diff
@@ -16,28 +17,49 @@ import androidx.navigation.ui.setupWithNavController
+                    val adapter = ArrayAdapter<Haushalt>(this, android.R.layout.simple_spinner_dropdown_item, haushaltItems)
+                    sp.adapter = adapter
+                }
+            })
```
Damit haben wir jetzt den Spinner und befüllen diesen jetzt mit den Daten aus der Datenbank oder?

Was wir noch machen müssten, ist dann in neine globale Variable vlt. den ajtuell ausgewählten haushalt zu schreiben um unnötige Datenbankzugriffe in den Fragmnents zu verhindern. Wäre das sinnvoll?

hoffmatteo replied:
Der Tobias hatte das doch so gelöst, dass wir uns den Spinner in die jeweiligen Fragments holen und durch den dann den ausgewählten Haushalt kriegen.

Extra Datenbankzugriffe brauchen wir dafür nicht, das wird ja durch LiveData automatisch geupdatet, aber nur in der MainActivity.

emuguy1 replied:
Ah ok. Verstehe. Wenn dann ein neuer Haushalt erzeugt wird, kommt der also dann auch automatisch in den Spinner?

hoffmatteo replied:
Ja, das wird ja durch den Observer automatisch gesetzt.

emuguy1 commented on app/src/main/java/com/example/stromtracker/database/Haushalt.kt:
```diff
@@ -70,6 +70,10 @@ data class Haushalt(
         name = n
     }
+
+    override fun toString():String {
+        return name
+    }
```
Warum überschreibst du hier die toString Methode?

hoffmatteo replied:
Für den Standard-Listadapter der Recyclerview, eventuell braucht man das nicht, aber so wie ich das recherchiert habe, ist es sinnvoll die toString Methode zu setzen.

emuguy1 replied:
Ah ok.

emuguy1 commented on app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt:
```diff
@@ -1,125 +1,186 @@
+                        geraeteViewModel.insertRaum(Raum("test", 1))
+                        geraeteViewModel.insertRaum(Raum("zet", 1))
                         geraeteViewModel.insertKategorie(Kategorie("test", null))
-                        geraeteViewModel.insertRaum(Raum("test"))
```
Das sollten wir auf jeden Fall noch in die MainActivity einfügen, da wir sonst eine Inkosistente Datenbank bekommen, wenn der User beim ersten Aufruf z.B. erst die Geräte Seite aufruft. Bzwe. ist hier das Problem, dass seöbst wenn ein Haushalt existiert, hier nochmal einer erstellt wird, und dann aber dem alten Haushalt die Geräte hinzugefügt werden.

Aber das ist etwas, dass wir noch in allen Fragments machen müssen, die auf die Datenbank zugreifen.

hoffmatteo replied:
Ist wie gesagt nur zum testen für mich, das ist nichts was bleiben soll.

emuguy1 replied:
Ja passt.

emuguy1 commented on app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt:
```diff
@@ -1,125 +1,186 @@
+                buttonSortName.paintFlags = 0
+
+            }
+            else -> {
+                //Toast.makeText(v.context, String.format(Locale.GERMAN,"%d was pressed.", v.id), Toast.LENGTH_SHORT).show()
+            }
```
Du solltest hier wenigstens ein Log ausgeben, dass irgendwas schie gelaufen ist, da das ja eigentlich nicht passieren sollte. Aber leer lassen, würde ich lassen

hoffmatteo replied:
Da kann ja nichts schief gehen an sich, wir wollen ja nur dass er bei bestimmten Button-Clicks was macht.

Wenn er keinen Button klickt, sondern irgendetwas anderes, dann soll doch keine Fehlermeldung o.ä kommen, da soll ja nichts passieren.

emuguy1 replied:
Aber da kommt man jetzt doch gar nicht hin, außer irgendwas ist schiefgelaufen, oder verstehe ich das falsch?

hoffmatteo replied:
Theoretisch kommt man da nur hin wenn der OnClickListener noch auf etwas anderes als die Buttons gesetzt wurde, was ja gar nicht sein kann.

Aber an sich kann man das noch hinzufügen.

tobireith replied:
Da gebe ich Matteo recht, das else wird eigentlich nicht benötigt und somit auch der Toast nicht unbedingt.
Das else wird, wie Matteo schon gesagt hat, nur aufgerufen wenn auf etwas ein Klick-Listener gesetzt wurde und die ID bei dem switch case in der  OnClick Funktion nicht behandelt wurde. Es ist sozusagen der default case.

emuguy1 commented on app/src/main/res/layout/fragment_geraete_edit.xml:
```diff
@@ -0,0 +1,211 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/geraet_edit_textview_margin"
+            android:layout_marginTop="16dp"
```
Ich glaub da haben wir auch schon was in dimen. Wöäre gut, wenn du das auch noch in Dimen auslagerst.

hoffmatteo replied:
Ja, die Auslagerung in dimen muss ich noch weiter machen bzw. sind wir da wahrscheinlich auch auf unterschiedlichen Versionen der dimens Klasse, deswegen habe ich das erstmal nach hinten verschoben.

emuguy1 replied:
Ja das ist eine gute Sache. Muss ich ja auch noch machen

emuguy1 commented on app/src/main/res/layout/fragment_geraete_edit.xml:
```diff
@@ -0,0 +1,211 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginEnd="16dp"
+            android:layout_marginTop="35dp"
+            android:layout_marginRight="16dp"
+            android:layout_marginBottom="16dp"
```
Margin Top und Bottom sind bei den Buttons ja auch gleich, da wär auch dut, die dann in die dimen auszulagern, so dass wenn man dass bei einem ändert, dann alle dementsprechend verschoben werden.

emuguy1 commented on app/src/main/res/layout/fragment_geraete_edit_produzent.xml:
```diff
@@ -0,0 +1,169 @@
+                android:id="@+id/geraete_edit_produzent_save"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginTop="24dp"
```
Gleiche wie eins davor.

emuguy1 commented on app/src/main/res/layout/fragment_geraete_new_produzent.xml:
```diff
@@ -0,0 +1,149 @@
+            android:layout_marginTop="24dp"
+            android:layout_marginEnd="16dp"
+            android:layout_marginRight="16dp"
+            android:layout_marginBottom="16dp"
```
Und hier ist es dann auch nochmal.

Muss ich aber auch noch machen  : )

emuguy1 commented on app/src/main/res/layout/fragment_geraete_new_verbraucher.xml:
```diff
@@ -0,0 +1,190 @@
+            android:layout_height="wrap_content"
+            android:layout_marginEnd="16dp"
+            android:layout_marginTop="35dp"
+            android:layout_marginRight="16dp"
```
Hier auch noch

emuguy1 reviewed (approved):
Ist echt gut geworden. Wie gesagt sind bei den Dimens noch ein paar, die wir auslagern könnten. Da muss ich mir aber auch an die eigene Nase fassen, weil das bei mir auch noch so ist. 
Aber damit wir es auf dem Schirm haben.

hoffmatteo committed:
Insert default Haushalte, first try at Haushalt-Dividing

tobireith commented on app/src/main/java/com/example/stromtracker/ui/geraete/geraet_new/GeraeteNewVerbraucherFragment.kt:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    override fun onNothingSelected(parent: AdapterView<*>?) {
+        TODO("Not yet implemented")
```
Ich glaube das TODO kann man dann noch löschen, oder willst du da noch etwas hinzufügen?

hoffmatteo replied:
Nein, das TODO wurde automatisch hinzugefügt, das kann weg.

tobireith commented on app/src/main/java/com/example/stromtracker/ui/geraete/geraet_edit/GeraeteEditVerbraucherFragment.kt:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    override fun onNothingSelected(parent: AdapterView<*>?) {
+        TODO("Not yet implemented")
```
Auch dieses TODO kann man denke ich noch löschen

hoffmatteo replied:
Ja, siehe oben

tobireith committed:
Energieproduzenten werden nun in eigener Liste angezeigt. Es wird zwischen Produzent und Verbraucher bei Edit unterschieden.

TODO: Floating Action Button plazieren

tobireith committed:
Bugfix bei ProduzentEdit, Recyclerview hat nun marginBottom damit der FAB Platz hat

hoffmatteo committed:
First try at using livedata to update currentHaushalt, contains errors

tobireith committed:
FAB Menu ist nun nichtmehr klickable, Warning fixes im Layout

FAB Menu ist nun nichtmehr klickable, damit blockiert es nicht mehr Klicks auf Sachen hinter dem Menü.

tobireith committed:
Merge branch 'geraete' into geraete_produzent

hoffmatteo committed:
Continued livedata linking, still doesn't work

hoffmatteo committed:
Revert "Continued livedata linking, still doesn't work"

This reverts commit 9885b3f961c41d4aeeb4eb99096ce1b836496e8f.

Revert accidental commit to merge request, to restore V1 of geraete"

hoffmatteo committed:
Revert "First try at using livedata to update currentHaushalt, contains errors"

This reverts commit b891f523e7f88a559fc6807fc3fc20b84bf58503.

Second revert to reset geraete to V1

tobireith committed:
Merge branch 'geraete' into geraete_produzent

# Conflicts:
#	app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt

tobireith committed:
Revert "Continued livedata linking, still doesn't work"

This reverts commit 9885b3f961c41d4aeeb4eb99096ce1b836496e8f.

Reverting 9885b3f961c41d4aeeb4eb99096ce1b836496e8f

tobireith committed:
Revert "First try at using livedata to update currentHaushalt, contains errors"

This reverts commit b891f523e7f88a559fc6807fc3fc20b84bf58503.

Reverting b891f523e7f88a559fc6807fc3fc20b84bf58503

tobireith committed:
Merge pull request #21 from emuguy1/geraete_produzent

Geraete Produzent into Geraete - Anzeige der Produzenten

andreaseckner commented on app/build.gradle:
```diff
@@ -66,4 +66,7 @@ dependencies {
+    //Floatin Action Menu
+    implementation 'com.getbase:floatingactionbutton:1.10.1'
+
 }
```
It's best practise to have a newline at the end of each file. 🐱

hoffmatteo replied:
Will be added with the next commit, thank you.

andreaseckner commented on app/schemas/com.example.stromtracker.database.AppDatabase/1.json:
```diff
@@ -248,14 +266,35 @@
+      "INSERT OR REPLACE INTO room_master_table (id,identity_hash) VALUES(42, 'fb7c6408a51d0c05742a94cd166afd4b')"
     ]
   }
 }
```
No newline at end of file.

andreaseckner commented on app/src/main/java/com/example/stromtracker/database/Kategorie.kt:
```diff
@@ -36,4 +36,8 @@ data class Kategorie(
+        return name
+    }
+
 }
```
No newline at end of file.

andreaseckner commented on app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt:
```diff
@@ -1,125 +1,270 @@
-    }
-}
\ No newline at end of file
+    }
```
No newline at end of file.

andreaseckner commented on app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteListAdapter.kt:
```diff
@@ -0,0 +1,74 @@
+            }
+        }
+    }
+}
```
No newline at end of file.

andreaseckner commented on app/src/main/java/com/example/stromtracker/ui/geraete/geraet_edit/GeraeteEditProduzentFragment.kt:
```diff
@@ -0,0 +1,166 @@
+        }
+
+    }
+}
```
No newline at end of file.

andreaseckner commented on app/src/main/java/com/example/stromtracker/ui/geraete/geraet_new/GeraeteNewVerbraucherFragment.kt:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        }
+    }
```
No newline at end of file.

andreaseckner commented on app/src/main/res/drawable/fab_label_background.xml:
```diff
@@ -0,0 +1,11 @@
+        android:bottom="4dp"/>
+    <corners
+        android:radius="2dp"/>
+</shape>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/layout/fragment_geraete.xml:
```diff
@@ -1,23 +1,291 @@
+
+    </com.getbase.floatingactionbutton.FloatingActionsMenu>
+
 </androidx.constraintlayout.widget.ConstraintLayout>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/values/styles.xml:
```diff
@@ -16,4 +16,10 @@
+        <item name="android:textColor">@color/white</item>
+    </style>
+
 </resources>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/values/strings.xml:
```diff
@@ -65,5 +65,29 @@
+
+
 
 </resources>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/values/dimens.xml:
```diff
@@ -11,4 +11,16 @@
+    <dimen name="geraet_edit_textview_margin">24dp</dimen>
+    <dimen name="geraet_edit_editText_margin">8dp</dimen>
+
 </resources>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/values/colors.xml:
```diff
@@ -8,4 +8,10 @@
+    <color name="white">#fafafa</color>
+    <color name="white_pressed">#f1f1f1</color>
+    <color name="black_semi_transparent">#B2000000</color>
 </resources>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/layout/fragment_geraete_textview.xml:
```diff
@@ -0,0 +1,79 @@
+    </androidx.cardview.widget.CardView>
+
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
No newline at end of file.

andreaseckner commented on app/src/main/res/layout/fragment_geraete_edit_produzent.xml:
```diff
@@ -0,0 +1,169 @@
+
+
+        </androidx.constraintlayout.widget.ConstraintLayout>
+    </ScrollView>
```
No newline at end of file.

andreaseckner commented on app/src/main/java/com/example/stromtracker/MainActivity.kt:
```diff
@@ -16,28 +17,51 @@ import androidx.navigation.ui.setupWithNavController
 class MainActivity : AppCompatActivity() {
 
     private lateinit var appBarConfiguration: AppBarConfiguration
+    private lateinit var sp:Spinner
```
Missing space.

andreaseckner commented on app/src/main/java/com/example/stromtracker/MainActivity.kt:
```diff
@@ -16,28 +17,51 @@ import androidx.navigation.ui.setupWithNavController
+        var geraeteViewModel: GeraeteViewModel =  ViewModelProviders.of(this).get(GeraeteViewModel::class.java)
+        val haushaltItems:ArrayList<Haushalt> = ArrayList()
+
+
```
That's a lot blank lines.

andreaseckner commented on app/src/main/res/layout/fragment_geraete_new_produzent.xml:
```diff
@@ -0,0 +1,149 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="24dp"
+            android:text="@string/geraet_new_TextRaum"
```
snake_case and camelCase should not be mixed. The convention is to use snake_case within XML.

https://medium.com/mindorks/android-resource-naming-convention-42e4e8026614

hoffmatteo replied:
Will also be fixed in the next commit.

andreaseckner reviewed (approved):
Looks good. There are only minor things which is especially code formatting. The Android Studio code formatter or detekt could be used to resolve this automatically. If not, just make sure not to use to many blank lines within the code but end every file with a newline.

hoffmatteo committed:
Fixed naming convention + newline (except Produzent)

hoffmatteo committed:
Small fixes v2

hoffmatteo committed:
Merge branch 'develop' into geraete
